File: medical-app-mailer/models/conversationModel.js
const mongoose = require("mongoose");

// Schema for individual messages
const messageSchema = new mongoose.Schema(
  {
    sender: {
      type: mongoose.Schema.ObjectId,
      ref: "User",
      required: [true, "Un message doit avoir un expéditeur"],
    },
    content: {
      type: String,
      trim: true,
    },
    timestamp: {
      type: Date,
      default: Date.now,
    },
    readBy: [
      {
        type: mongoose.Schema.ObjectId,
        ref: "User",
      },
    ],
    type: {
      type: String,
      enum: ["text", "image", "file"],
      default: "text",
    },
    fileUrl: {
      type: String,
    },
    fileName: {
      type: String,
    },
    fileSize: {
      type: Number,
    },
    fileMimeType: {
      type: String,
    },
    status: {
      type: String,
      enum: ["sent", "delivered", "read"],
      default: "sent",
    },
  },
  {
    toJSON: { virtuals: true },
    toObject: { virtuals: true },
  }
);

// Schema for conversations
const conversationSchema = new mongoose.Schema(
  {
    patientId: {
      type: mongoose.Schema.ObjectId,
      ref: "User",
      required: [true, "Une conversation doit avoir un patient"],
    },
    doctorId: {
      type: mongoose.Schema.ObjectId,
      ref: "User",
      required: [true, "Une conversation doit avoir un médecin"],
    },
    patientName: {
      type: String,
      required: [true, "Le nom du patient est requis"],
    },
    doctorName: {
      type: String,
      required: [true, "Le nom du médecin est requis"],
    },
    lastMessage: {
      type: String,
      default: "",
    },
    lastMessageType: {
      type: String,
      enum: ["text", "image", "file"],
      default: "text",
    },
    lastMessageTime: {
      type: Date,
      default: Date.now,
    },
    lastMessageSenderId: {
      type: mongoose.Schema.ObjectId,
      ref: "User",
    },
    lastMessageReadBy: [
      {
        type: mongoose.Schema.ObjectId,
        ref: "User",
      },
    ],
    lastMessageUrl: {
      type: String,
    },
    isActive: {
      type: Boolean,
      default: true,
    },
  },
  {
    timestamps: true,
    toJSON: { virtuals: true },
    toObject: { virtuals: true },
  }
);

// Add index for faster queries
conversationSchema.index(
  { patientId: 1, doctorId: 1 },
  { unique: true }
);

// Static method to parse date strings
conversationSchema.statics.parseDateTime = function (dateTimeString) {
  if (!dateTimeString) return new Date();

  try {
    return new Date(dateTimeString);
  } catch (error) {
    console.error("Error parsing date:", error);
    return new Date();
  }
};

// Virtual populate messages
conversationSchema.virtual("messages", {
  ref: "Message",
  foreignField: "conversation",
  localField: "_id",
});

const Conversation = mongoose.model(
  "Conversation",
  conversationSchema
);
const Message = mongoose.model("Message", messageSchema);

module.exports = { Conversation, Message };

File: medical_app/lib/features/messagerie/data/models/conversation_model.dart
import 'package:medical_app/features/messagerie/domain/entities/conversation_entity.dart';

class ConversationModel extends ConversationEntity {
  ConversationModel({
    String? id,
    required String patientId,
    required String doctorId,
    required String patientName,
    required String doctorName,
    String lastMessage = '',
    String lastMessageType = 'text',
    required DateTime lastMessageTime,
    String? lastMessageSenderId,
    List<String>? lastMessageReadBy,
    String? lastMessageUrl,
    bool isActive = true,
    required DateTime createdAt,
    required DateTime updatedAt,
  }) : super(
         id: id,
         patientId: patientId,
         doctorId: doctorId,
         patientName: patientName,
         doctorName: doctorName,
         lastMessage: lastMessage,
         lastMessageType: lastMessageType,
         lastMessageTime: lastMessageTime,
         lastMessageSenderId: lastMessageSenderId,
         lastMessageReadBy: lastMessageReadBy ?? [],
         lastMessageUrl: lastMessageUrl,
         isActive: isActive,
         createdAt: createdAt,
         updatedAt: updatedAt,
       );

  factory ConversationModel.fromJson(Map<String, dynamic> json) {
    return ConversationModel(
      id: json['_id'] as String?,
      patientId: json['patientId'] as String? ?? '',
      doctorId: json['doctorId'] as String? ?? '',
      patientName: json['patientName'] as String? ?? 'Unknown Patient',
      doctorName: json['doctorName'] as String? ?? 'Unknown Doctor',
      lastMessage: json['lastMessage'] as String? ?? '',
      lastMessageType: json['lastMessageType'] as String? ?? 'text',
      lastMessageTime: parseDateTime(json['lastMessageTime']),
      lastMessageSenderId: json['lastMessageSenderId'] as String?,
      lastMessageReadBy:
          json['lastMessageReadBy'] != null
              ? List<String>.from(json['lastMessageReadBy'] as List)
              : [],
      lastMessageUrl: json['lastMessageUrl'] as String?,
      isActive: json['isActive'] as bool? ?? true,
      createdAt: parseDateTime(json['createdAt']),
      updatedAt: parseDateTime(json['updatedAt']),
    );
  }

  static DateTime parseDateTime(dynamic dateValue) {
    if (dateValue == null) {
      return DateTime.now();
    }

    if (dateValue is String) {
      try {
        return DateTime.parse(dateValue);
      } catch (e) {
        return DateTime.now();
      }
    } else if (dateValue is int) {
      try {
        return DateTime.fromMillisecondsSinceEpoch(dateValue);
      } catch (e) {
        return DateTime.now();
      }
    }

    return DateTime.now();
  }

  Map<String, dynamic> toJson() {
    final data = {
      'patientId': patientId,
      'doctorId': doctorId,
      'patientName': patientName,
      'doctorName': doctorName,
      'lastMessage': lastMessage,
      'lastMessageType': lastMessageType,
      'lastMessageTime': lastMessageTime.toIso8601String(),
      'lastMessageReadBy': lastMessageReadBy,
      'isActive': isActive,
    };

    if (id != null) {
      data['_id'] = id!;
    }
    if (lastMessageSenderId != null) {
      data['lastMessageSenderId'] = lastMessageSenderId!;
    }
    if (lastMessageUrl != null) {
      data['lastMessageUrl'] = lastMessageUrl!;
    }

    return data;
  }
}
